SwrveSDK/src/main/java/com/swrve/sdk/messaging/view/SwrveMessageView.java

```diff
@@ -1,250 +1,341 @@
-    }
+    // Bitmap cache
+    protected Set<WeakReference<Bitmap>> bitmapCache;
 
-    public SwrveMessageView(Context context, SwrveMessage message, SwrveMessageFormat format, ISwrveInstallButtonListener installButtonListener,
-                            ISwrveCustomButtonListener customButtonListener, boolean firstTime, int rotation, int minSampleSize) throws SwrveMessageViewBuildException {
+    public SwrveMessageView(Context context, Messenger messenger, SwrveMessage message, SwrveMessageFormat format, int minSampleSize) throws SwrveMessageViewBuildException {
         super(context);
-        this.message = message;
+        this.messenger = messenger;
         this.format = format;
-        this.firstTime = firstTime;
-        initializeLayout(context, message, format, installButtonListener, customButtonListener, rotation, minSampleSize);
-    }
-
-    protected void initializeLayout(final Context context, final SwrveMessage message, final SwrveMessageFormat format, ISwrveInstallButtonListener installButtonListener,
-                                    ISwrveCustomButtonListener customButtonListener, int rotation, int minSampleSize) throws SwrveMessageViewBuildException {
-        innerMessageView = new SwrveInnerMessageView(context, this, message, format, installButtonListener, customButtonListener, minSampleSize);
-        setBackgroundColor(format.getBackgroundColor());
-        setGravity(Gravity.CENTER);
-        setLayoutParams(new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT));
-        addView(innerMessageView);
-        setClipChildren(false);
-
-        if (rotation != 0) {
-            // Flip internal message so that it can still fit in the screen
-            RotateAnimation rotate = new RotateAnimation(0, rotation, Animation.RELATIVE_TO_SELF, 0.5f, Animation.RELATIVE_TO_SELF, 0.5f);
-            rotate.setFillAfter(true);
-            innerMessageView.startAnimation(rotate);
+        // Sample size has to be a power of two or 1
+        if (minSampleSize > 0 && (minSampleSize % 2) == 0) {
+            this.minSampleSize = minSampleSize;
         }
+        initializeLayout(context, message, format);
     }
 
-    /**
-     * @return the related message format.
-     */
-    public SwrveMessageFormat getFormat() {
-        return format;
-    }
-
-    /**
-     * @return the show animation.
-     */
-    public int getShowAnimation() {
-        return showAnimation;
+    public void notifyOfImpression(SwrveMessage message, SwrveMessageFormat format) {
+        Bundle bundle = new Bundle();
+        bundle.putInt("formatIndex", message.getFormats().indexOf(format));
+        sendMessage("impression", bundle);
     }
 
-    /**
-     * Set the message show animation.
-     *
-     * @param showAnimation
-     */
-    public void setShowAnimation(int showAnimation) {
-        this.showAnimation = showAnimation;
+    private void notifyOfInstallButtonPress(SwrveButton button, String appInstallLink) {
```

Comment:
These notify methods belong in their own class as they have their own set of logic and rules which can be far easier unit tested as a separate thing.